From: "Kasireddy, Vivek" <vivek.kasireddy@intel.com>
Subject: RE: [fs?] [mm?] INFO: task hung in remove_inode_hugepages
Date:   Mon Jul 8 23:04:48 2024 -0700


return -EINVAL if the end offset is greater than the size of memfd

Link: https://lkml.kernel.org/r/IA0PR11MB71850525CBC7D541CAB45DF1F8DB2@IA0PR11MB7185.namprd11.prod.outlook.com
Signed-off-by: Vivek Kasireddy <vivek.kasireddy@intel.com>
Reported-by: <syzbot+f1d7fb4f94764243d23e@syzkaller.appspotmail.com>
Testted-by: <syzbot+f1d7fb4f94764243d23e@syzkaller.appspotmail.com>
Cc: Dave Airlie <airlied@redhat.com>
Cc: Gerd Hoffmann <kraxel@redhat.com>
Cc: Muchun Song <muchun.song@linux.dev>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/gup.c |    3 +++
 1 file changed, 3 insertions(+)

--- a/mm/gup.c~mm-gup-introduce-memfd_pin_folios-for-pinning-memfd-folios-fix-2
+++ a/mm/gup.c
@@ -3814,6 +3814,9 @@ long memfd_pin_folios(struct file *memfd
 	if (!shmem_file(memfd) && !is_file_hugepages(memfd))
 		return -EINVAL;
 
+	if (end >= i_size_read(file_inode(memfd)))
+		return -EINVAL;
+
 	if (is_file_hugepages(memfd)) {
 		h = hstate_file(memfd);
 		pgshift = huge_page_shift(h);
_
